File: subset sum equal to target/SpaceOptimization.java

class SpaceOptimization {
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4};
        int target = 4;

        int size = arr.length;

        boolean[] prev = new boolean[target + 1];
        prev[0] = true;

        for (int i = 1; i < size; i++) {

            boolean[] cur = new boolean[target + 1];
            cur[0] = true;

            for (int j = 1; j < target + 1; j++) {
                boolean nonTake = prev[j];
                boolean take = false;
                if (j >= arr[i]) {
                    take =  prev[j - arr[i] ] ;
                }

                cur[j] = (nonTake || take);
            }

            prev =  cur;
        }


        System.out.println(prev[target]);
    }
}

File: subset sum equal to target/SpaceOptimization.py
if __name__ == '__main__':
    arr = [1,2,3,4]
    target_sum = 4

    size = len(arr)
    prev = [ False for _ in range(target_sum+1)]
    prev[0] = True 

    for index in range(1,size):

        cur = [ False for _ in range(target_sum+1)]
        cur[0] = True 

        for target in range(1,target_sum+1):
            nonTake = prev[target]
            take = False
            if target >= arr[index] :take = prev[target - arr[index]]

            cur[target] = nonTake or take

        prev = cur
    
    print(prev[target_sum])